keps/1337-reclaim-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,373 @@
+# KEP-1337: Reclaim within cohort while borrowing
+
+<!--
+This is the title of your KEP. Keep it short, simple, and descriptive. A good
+title can help communicate what the KEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a KEP and for
+highlighting any additional information provided beyond the standard KEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Flopping between preempting pair of workloads](#flopping-between-preempting-pair-of-workloads)
+- [Design Details](#design-details)
+  - [API](#api)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Beta](#beta)
+    - [GA](#ga)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Minimal API](#minimal-api)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces ClusterQueue-level API to support workload preemption
+to reclaim quota within cohort for a workload that requires borrowing.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Currently, preemption can only happen within ClusterQueue, or within cohort, but
+only if the workload that wants to be submitted does not require borrowing.
+This requirement is too constraining for some setups.
+
```

Comment:
Can you add why we need to extend API, i.e. why modifying scheduler won't work?

Reply:
Yes, the main motivation is the ability to Disable it (setting Never). Users may want to disable. When it is enabled, it might be that there are workloads for which we want to attempt preemption, but because there are no lower priority workloads, the workload would keep blocking the queue. This is similar reason, for which we allow users to disable preemption when within ClusterQueue, or within cohort (but without borrowing.